Define schemas for ABI and Network Objects

gnidan opened the pull request:
Ref: #5 

Following https://github.com/ethereum/wiki/wiki/Ethereum-Contract-ABI#json as reference

gnidan committed:
Define ABI schema and separate from contract schema

gnidan committed:
Rename spec to contract-object.spec.json

gnidan commented on test/schema.js:
```diff
@@ -1,33 +1,11 @@
-    var valid = validator(MetaCoin);
-    assert.ifError(validator.errors);
-  });
-
```
These tests are all pointless. The other tests will rightly explode if the schema's invalid

gnidan commented:
Sorry @tcoulter decided I want to do a bit more with this before calling it done

gnidan committed:
Add some whitespace for schema readability

gnidan committed:
Make abi.spec.json a proper root-level obj schema

gnidan committed:
Remove some unnecessary blank newlines

gnidan committed:
Separate NetworkObject into own spec file

gnidan committed:
Disallow additional properties in network object

gnidan committed:
Add validate option to normalize

gnidan committed:
Use schema defaults to fill in object

gnidan committed:
Add tests for ABI schema and fix issues

tcoulter commented on index.js:
```diff
@@ -205,7 +211,11 @@ var TruffleContractSchema = {
     normalized.schemaVersion = pkgVersion;
 
-    return normalized;
+    if (options.validate) {
```
This is a nice feature. Are you using it anywhere outside of `truffle-contract-schema`? 

We should add a test for it. A good test for this could be one that has an invalid ABI, and therefore `normalize` would error if `validate == true` (meaning it must have called `validate()` in order to error).

gnidan replied:
> Are you using it anywhere outside of truffle-contract-schema?

Not yet - just figured it'd be a good option to have. I tried running various tests in the suite with `validate` turned on, but there were errors with different parts of Truffle producing barely invalid ABIs

tcoulter replied:
> Truffle producing barely invalid ABIs

Truffle's tests, correct? Truffle itself only "produces" abis that come directly from Solidity (using my interpretation of the word "produces").

gnidan replied:
Yep, Truffle's tests.

Oh, I remember now. The problem was with link references not conforming to the regular expression. Seemed like a bit too much trouble to fix right away, because it'd require deciding to include link references as part of the schema (and I know you have plans for those)

tcoulter replied:
When you say "link references", so you mean the `linkReferences` key? If not, in general, the way we perform links now *should* go in the schema - that is, network objects's `links` key should be validated to have the right structure.

Remember: The new schema is the authority on what is and isn't correct, so if the tests are sending in bad links we should change the tests (on `develop`, of course).

gnidan replied:
I mean the link references inside bytecode... because solc puts path delimiter characters inside the underscore block, and you've talked about standardizing what that looks like.

I would think the correct answer is to perform normalization of the unlinked bytecode itself, but not sure when we should do that.

tcoulter replied:
Ah. Hmm, I'd make the bytecode regex less strict, at least for now. Effectively because link references can contain any character in between those underscores, at the moment. You could also write a regex that takes into account the underscores and the ability to add any character between those underscores (and that the reference length should be 40 characters total), but that sounds like too much work for a single regex. I'm fine with either option at this point, assuming you don't fight with the regex for a long time.

tcoulter commented on index.js:
```diff
@@ -146,18 +149,21 @@ var TruffleContractSchema = {
-    var validate = ajv.compile(contractSchema);
-    if (validate(contractObj)) {
+    var ajv = new Ajv({ useDefaults: true });
+    ajv.addSchema(abiSchema);
```
For my edification: How does this know that `contractObj.abi` is what should be validated with the abiSchema, vs. the `contractObj` itself?

gnidan replied:
Ah, the magic of json-schema! Specifically it's [this `$ref`](https://github.com/trufflesuite/truffle-contract-schema/pull/7/files#diff-76dc6ee27dce76d47ec972f216f2b8cbR8) that does it.

tcoulter reviewed (commented):
Nearly LGTM. Couple questions:

gnidan commented:
I have a bit more work to push up locally on this, then I will look at writing tests for the validate option.

gnidan committed:
Iterate on ABI schema

- Pull out name/type as own definitions.
- Build regex for valid ABI types
- Add title
- Ensure ABI items don't match more than one event/function schema

gnidan committed:
Add descriptions and examples to properties

gnidan committed:
Increase flexibility for Name/Bytecode

- ABI names can be blank in some cases
- Bytecode should not stipulate anything about the 38 characters after `__`
  in refs

gnidan commented:
Made a ticket for tracking the work of internal schema validation within truffle tests: https://github.com/trufflesuite/truffle-contract-schema/issues/9

gnidan committed:
Add schema docs for contract/network objects

gnidan committed:
Remove unnecessary Network Object: prefix in ref

tcoulter committed:
Merge pull request #10 from trufflesuite/readme

Add schema docs for contract/network objects
